src/Bicep.Core/Parsing/ParamsParser.cs

```diff
@@ -0,0 +1,1139 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Globalization;
+using System.Linq;
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Extensions;
+using Bicep.Core.Navigation;
+using Bicep.Core.Syntax;
+
+namespace Bicep.Core.Parsing
+{
+    public class ParamsParser
+    {
+        private readonly TokenReader reader;
+
+        private readonly ImmutableArray<IDiagnostic> lexerDiagnostics;
+
+        public ParamsParser(string text)
+        {
+            // treating the lexer as an implementation detail of the parser
+            var diagnosticWriter = ToListDiagnosticWriter.Create();
+            var lexer = new Lexer(new SlidingTextWindow(text), diagnosticWriter);
+            lexer.Lex();
+
+            this.lexerDiagnostics = diagnosticWriter.GetDiagnostics().ToImmutableArray();
+
+            this.reader = new TokenReader(lexer.GetTokens());
+        }
+
+        public ProgramSyntax Program()
+        {
+            var declarationsOrTokens = new List<SyntaxBase>();
+
+            while (!this.IsAtEnd())
+            {
+                // this produces either a declaration node, skipped tokens node or just a token
+                var declarationOrToken = Statement();
+                declarationsOrTokens.Add(declarationOrToken);
+
+                // if skipped node is returned above, the newline is not consumed
+                // if newline token is returned, we must not expect another (could be a beginning of a declaration)
+                if (declarationOrToken is ITopLevelDeclarationSyntax)
+                {
```

Comment:
The check won't work since neither `ParameterAssignmentSyntax` nor `UsingDeclarationSyntax` implements `ITopLevelDeclarationSyntax`. We may need to change it to `StatementSyntax`.